# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/AsyncCanvasRenderer.cpp
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 gfx/layers/AsyncCanvasRenderer.cpp
--- a/gfx/layers/AsyncCanvasRenderer.cpp	Wed Jan 08 21:52:03 2020 +0000
+++ b/gfx/layers/AsyncCanvasRenderer.cpp	Wed Jan 08 22:19:14 2020 +0000
@@ -11,10 +11,13 @@
 #include "GLReadTexImageHelper.h"
 #include "GLScreenBuffer.h"
 #include "mozilla/dom/HTMLCanvasElement.h"
+#include "nsICanvasRenderingContextInternal.h"
 #include "mozilla/layers/BufferTexture.h"
 #include "mozilla/layers/CanvasClient.h"
+#include "mozilla/layers/CompositableForwarder.h"
 #include "mozilla/layers/TextureClient.h"
 #include "mozilla/layers/TextureClientSharedSurface.h"
+#include "mozilla/layers/ShadowLayers.h"
 #include "mozilla/ReentrantMonitor.h"
 #include "nsIRunnable.h"
 #include "nsThreadUtils.h"
@@ -93,6 +96,18 @@
   mCanvasClient = aClient;
   if (aClient) {
     mCanvasClientAsyncHandle = aClient->GetAsyncHandle();
+    if (mContext) {
+      MOZ_ASSERT(mCanvasClient->GetForwarder() &&
+                 mCanvasClient->GetForwarder()->AsLayerForwarder() &&
+                 mCanvasClient->GetForwarder()
+                     ->AsLayerForwarder()
+                     ->GetShadowManager());
+      LayerTransactionChild* ltc =
+          mCanvasClient->GetForwarder()->AsLayerForwarder()->GetShadowManager();
+      DebugOnly<bool> success =
+          mContext->UpdateCompositableHandle(ltc, mCanvasClientAsyncHandle);
+      MOZ_ASSERT(success);
+    }
   } else {
     mCanvasClientAsyncHandle = CompositableHandle();
   }